pkg/resource/deploy/snapshot.go

```diff
@@ -55,49 +55,54 @@ func NewSnapshot(manifest Manifest, secretsManager secrets.Manager,
 // references which do not need to be indirected through any alias lookups, and which instead refer directly to the URN
 // of a resource in the resources map.
 //
-// Note: This method modifies the snapshot (and resource.States in the snapshot) in-place.
-func (snap *Snapshot) NormalizeURNReferences() error {
-	if snap != nil {
-		aliased := make(map[resource.URN]resource.URN)
-		fixUrn := func(urn resource.URN) resource.URN {
-			if newUrn, has := aliased[urn]; has {
-				return newUrn
+// Note: This method does not modify the snapshot (and resource.States
+// in the snapshot) in-place, but returns an independent structure,
+// with minimal copying necessary.
+func (snap *Snapshot) NormalizeURNReferences() (*Snapshot, error) {
+	if snap == nil {
+		return nil, nil
+	}
+
+	aliased := make(map[resource.URN]resource.URN)
```

Comment:
In the old version, construction of `aliased` table was interleaved with traversing `snap.Resources`. I worry if this was intentional or has some subtle consequences, compared to precomputing the table in a single pass here, and then proceeding to another pass.

Reply:
I was also curious about the construction of the alias map in the loop. I suspect that the order of resources is determined by the topological sort (or reverse as the case may be), but I didn't see an obvious constraint.

This change seems more resilient.